### ISSUE TITLE
Add `accessibleName` and `helpText`

### ISSUE THREAD

**Issue by pixelzoom** [phetsims/models-of-the-hydrogen-atom#67]
In 9/25/24 standup meeting, @DianaTavares mentioned that she has been adding `accessibleName` and `helpText` attributes to the [sim design document](https://docs.google.com/document/d/1fZT_vDD8sX8nkSpTxHOyoP6-XrAIJYo44dacLIbZO14/edit?pli=1). She is doing it for practice; there are no plans to implement this for the 1.0 release. 

@jbphet suggested creating this issue so that future developers/designers will know that this work had been started.

**Comment by pixelzoom** [comment]
Related to my estimate in #71, @kathy-phet asked us to proceed with the `accessibleName` part of this issue, so removing the deferred label.  Kathy's instructions in Slack#models-of-the-hydrogen-atom were:

> @DianaTavares - You had previously asked, I believe, about doing accessibleName in MOTHA, and with Chris' assessment, I think we can proceed within budget and you @Nancy-Salpepi.  @terracoda and @arouinfar can work on coming up with the accessibleNames strings for this sim. Would you have time to focus on meeting to design those names this coming iteration? (which starts on Thursday)? Please keep track of how much design time it takes, so we can get a sense of that as well as implementation time.

**Comment by DianaTavares** [comment]
We finish with the design for accessible names and help Text in the design doc. I add them in the design doc, as it was proposed by the accessibility team, in the ["iteravitity"](https://docs.google.com/document/d/1fZT_vDD8sX8nkSpTxHOyoP6-XrAIJYo44dacLIbZO14/edit#heading=h.n6bxa5eascev) section of each screen. 

I think that plan this in the design doc is good! But we need a meeting to communicate to @pixelzoom how to convert the content in the document to code for the a11y view. I am not sure if is better to meet soon or until spectrometer and the second screen have more development,

**Comment by pixelzoom** [comment]
Thanks @DianaTavares.  It would be best to add these features incrementally, as the sim is implemented -- there's no need to wait until the sim is feature-complete.  With that in mind, I'll review the design document, then schedule a meeting for next week to review how to implement what is specified.

**Comment by arouinfar** [comment]
@DianaTavares @pixelzoom I added a few suggestions/clarifying comments to the doc.

**Comment by pixelzoom** [comment]
As part of this work, @kathy-phet also requested that I "work with dev team around strategy for translation of these (that would scale for more complex strings)."

**Comment by pixelzoom** [comment]
I plan to start with the approach used in the past.  That is, put all of these strings under `ModelsOfTheHydrogenAtomStrings.a11y`, where they are not translatable.  This will allow the designers to test/review the strings and their behavior.

Then I'll start a disussion with devs about how to make these translatable and scalable.

**Comment by pixelzoom** [comment]
@arouinfar pointed me to ConcentrationControlModeRadioButtonGroup in GreenhouseGasConcentrationPanel.ts as an example of how accessibleName and helpText are done for radio button groups.  MOTHA radio button groups would probably be `labelTagName: 'h3'`.

```ts
        // pdom
        labelTagName: 'h4',
        labelContent: GreenhouseEffectStrings.a11y.concentrationPanel.experimentModeStringProperty,
        helpText: GreenhouseEffectStrings.a11y.concentrationPanel.experimentModeHelpTextStringProperty,
```

Questions:
* Why doesn't RectangularRadioButtonGroup (and other groups?) have `accessibleName`?
* Why are we using head tags, which are very DOM dependent, easily broken, brittle, in this way?

**Comment by pixelzoom** [comment]
I implemented most of the specification for accessibleName and helpText, and added some questions to the design document.  We had a design meeting this morning: @DianaTavares @arouinfar @terracoda @pixelzoom.  Many things about the design were revised, and we identified some general issues that need attention in common code.

## Sim-specific change requests:

- [ ] Light Mode => Light Source Mode

- [ ] Add helpText for Light Source Mode

- [ ] Absorption/Emission => Absorption and Emission

- [ ] Show absorption/emission dialog for direct access to wavelengths. => Show Absorption and Emission dialog for electron state transitions and their associated wavelengths.

- [ ] Absorption / Emission Dialog => Absorption and Emission Dialog

- [ ] Remove helpText for Absorption/Emission dialog.

- [ ] Add structure around "Wavelength {{wavelength}} nanometers" buttons. Group under a parent Node?
```ts
accessibleName: Transition Wavelengths
helpText: Set the monochromatic light source to specific wavelength.
```

- [ ] "Wavelength {{wavelength}} nanometers" => "{{wavelength}} nanometers"

- [ ] Experiment Model => Experiment or Model

- [ ] experimentOrModelHelpText => Collect experimental data or explore atomic models. 

- [ ] Bohr => Billiard Ball typo for first radio button (copy-paste error)

- [ ] missing atomicModelHelpText: Choose Atomic Model to explore.

- [ ] Investigate why content is above Spectrometer accordion box expand/collapse button. See also RAP MyChallengeAccordionBox.

- [ ] Take up to four snapshots of the spectrometer data. => Take up to four snapshots of spectrometer data. 

- [ ] Delete Snapshot => Delete {{snapshotName}} Snapshot

- [ ] "My Snapshots" - how has this been done for other dialogs? Probably a heading and labelContent

- [ ] Info button helpText: Definition of the quantum numbers. => Show definition of quantum numbers.

- [ ] Quantum Numbers title for the dialog.

- [ ] Electron Energy Level helpText: Observe changes in the electron’s state as shown in an energy level diagram. (DL/AR/NS to refine.)

## Common-code issues to be created in GitHub:

- [ ] LaserPointerNode seems to be ignoring `buttonAccessibleName`. This patch seems to fix the problem, but needs to be tested to see if it impacts other usages.
```diff
- labelContent: options.buttonAccessibleName,
- labelTagName: 'label',
+ accessibleName: options.buttonAccessibleName,
```

- [ ] For AquaRadioButtonGroup and RectangularRadioButtonGroup, why does helpText appear after radio buttons in A11y View? E.g. light mode radio buttons in MOTHA

- [ ] NumberControl does not work with accessibleName and helpText.  A workaround of using `sliderOptions.accessibleName` has been used.  Additionally, Slider does not know about units, so we get “396” instead of “396 nm”.  NumberControl can certainly delegate to Slider, and configure Slider correctly.  How does ‘units: nm’ result in “381 nanometers” being read?

- [ ] Discuss with JG where helpText should appear by default for Dialog, with placement options.

- [ ] TimeControlNode accessibleName and helpText is not general, is incorrect with some options. "Timing" is incorrect, as is TimingControlsKeyboardHelpSection, recommended to change to "Time".

- [ ] EraseButton should have a default accessibleName "Erase".

- [ ] RectangularRadioButtonGroupItem is ignoring labelContent and using tandemName to create accessible name.X

- [ ] ABSwitch does not support for accessibleName and helpText: include TS & AR in design, JG or MK in dev (depending on who worked on OnOffSwitch)

- [ ] What is the pattern for AccordionBox?  It's not identified in AccordionBox.md. 

- [ ] Why is helpText for an AccordionBox shown in large bold font in A11y View? E.g. ElectronEnergyLevelAccordionBox.

**Comment by pixelzoom** [comment]
I've made all of the change requests listed in https://github.com/phetsims/models-of-the-hydrogen-atom/issues/67#issuecomment-2415033501.  Those that are blocked by common-code work are annotated with a ❗, followed by the GitHub issue for the common-code work.

I've created GitHub issues for all of the general things that we identified in https://github.com/phetsims/models-of-the-hydrogen-atom/issues/67#issuecomment-2415033501.

Do we review and continue to refine the MOTHA implementation of accessibleName and helpText? Or was this sufficient for identifying improvements needed in common code?  @DianaTavares Let's discuss next steps.

**Comment by pixelzoom** [comment]
> ... @DianaTavares @kathy-phet Let's discuss next steps.

Discussed with @kathy-phet.  I'm going to work on resolving the remaining change requests, with assistance from @jessegreenberg.  We'll also try to address some of the TODOs in the MOTHA code that are using less desirable ways of implementing description.  When we get that done, we'll have review of the work.

@kathy-phet also requested that we add a static scene description for reach screen.

**Comment by pixelzoom** [comment]
> ... @DianaTavares @kathy-phet Let's discuss next steps.

Discussed with @kathy-phet.  @jessegreenberg and I will work on improving/adding programming APIs for things that are used by MOTHA, address the TODOs in the MOTHA code that are using less desirable ways of implementing description.  I'll going to work on completing the remaining change requests.  I'll do some refinements with @DianaTavares and @arouinfar.  When things are in a good place, we'll loop @terracoda back in and have a team review of where we're at with this feature.

@kathy-phet also requested that we add a static scene description for reach screen. @jessegreenberg and I started some work related to that in https://github.com/phetsims/joist/issues/987.

**Comment by pixelzoom** [comment]
@jessegreenberg To make it easier to identify which common-code work to prioritize for MOTHA, I've added TODO comments to the code of the form "//TODO description:", followed by the common-code issue number and some verbage. There are currently 13 such comments. Please have a look, and let's coordinate on how to address these.

**Comment by pixelzoom** [comment]
@DianaTavares @arouinfar This is ready for you to have a look at the A11y View. 

Note that for the change requests in https://github.com/phetsims/models-of-the-hydrogen-atom/issues/67#issuecomment-2415033501, there are 2 requests that I'm still working on, noted by ❗.  They both require common-code work.

**Comment by DianaTavares** [comment]
I reviewed the a11y view, and well, this is my first time reviewing this type of implementation, then my comments are based on comparing this sim with others and what I expected to see. My observations are:

1) The order of the controls when the spectrometer is open is strange. But I see that this issue is marked by ❗

2) Some implementations need the sim to continue with the features, for example, the name of the snapshots 
![image](https://github.com/user-attachments/assets/1adf7d4c-1093-4b27-b804-865829dee454)

3) I am almost sure that the visual of the text is not related to the goal of the accessible names, but if it is possible to organize this section, will be better! 

![image](https://github.com/user-attachments/assets/090daeda-4bce-4946-997b-7292324c895e)

The experiment button is very close to the wavelength values, and it will be better if there is just at least a space between those buttons. 

4) The second screen is hard to review because it is not fully implemented. 

The rest is what I expected to see!

**Comment by pixelzoom** [comment]
Thanks for the feedback @DianaTavares.

> 1. The order of the controls when the spectrometer is open is strange.  ...

This has been fixed in https://github.com/phetsims/sun/issues/902.

> 3. I am almost sure that the visual of the text is not related to the goal of the accessible names, but if it is possible to organize this section, will be better!

Formatting and organization of the A11y View is not the responsibility of the sim.  And according to @jessegreenberg, the current implementation of A11y View is rather primitive, and just displays what is in the PDOM. I've opened https://github.com/phetsims/chipper/issues/1510 to start a discussion about what formatting is needed in A11y View to support design and development.

> 2. Some implementations need the sim to continue with the features, for example, the name of the snapshots.
> 4. The second screen is hard to review because it is not fully implemented.

Right. What you're currently seeing for the snapshots are placeholders, based on what was designed.  For the second screen, it's sufficient to review the Electron Energy Levels dialog description, since that's the only UI component that is diffeent in that screen; all others will be the same as the first screen.

**Comment by pixelzoom** [comment]
Status on common-code improvements related to MOTHA... @jessegreenberg has been doing a great job at cranking through these, and we're down to 4 `//TODO description` comments in MOTHA that identify places where the common-code API needs improvement.   These changes have been made without any changes to the description design, solely improvements to the programming API, making things easier for sim developers in the typical cases, while retaining flexibility.

**Comment by pixelzoom** [comment]
Sorry @arouinfar, I neglected to ask if you wanted to review change requests from 10/15 design meeting: https://github.com/phetsims/models-of-the-hydrogen-atom/issues/67#issuecomment-2415033501.  @DianaTavares has reviewed in https://github.com/phetsims/models-of-the-hydrogen-atom/issues/67#issuecomment-2443052757, and I responded in the comment immediately after.

**Comment by pixelzoom** [comment]
All blocking common-code work has been addressed by @jessegreenberg (thanks!) so I will unassign him as we are now in the review/refine phase.

**Comment by pixelzoom** [comment]
@DianaTavares @arouinfar @terracoda Related to the 10/15 design meeting... I’ve addressed and closed all of the outstanding comments in the Google design doc related to accessibleName helpText.  If you were on a comment thread, you should receive a Google notification.  Let me know if there are questions.

**Comment by arouinfar** [comment]
> Sorry @arouinfar, I neglected to ask if you wanted to review change requests from 10/15 design meeting

I reviewed on main and everything looks good to me!

**Comment by DianaTavares** [comment]
I already review the implementation and it looks good! 


Looks like only @terracoda is asked to review to close this, but no sure if for her is going to be easy without the spectrometer and the second screen. No sure if it is better to have her input now, or when the sim has all the features implemented.

**Comment by pixelzoom** [comment]
I'm planning to flesh out the 2nd screen next week.

**Comment by pixelzoom** [comment]
In the commits above, I fleshed out the _Energy Levels_ screen, and added correct names and numbering for the Spectrometer snapshots.   There is a lot of duplication in the ScreenView subclasses that I plan to address later.  But this address the immediate goal of making it easier to review description features.

**Comment by pixelzoom** [comment]
@terracoda This issue is ready for your review and feedback, including:

- [ ] Does accessible name, help text, and focus order match the specification in the Interactivity sections of the design document?
- [ ] Are there any additional change requests?
- [ ] Anything else for this issue?

**Comment by pixelzoom** [comment]
In Slack#motha...  Help text for screen buttons (on Home screen and in navigation bar) was identified as missing.  Tracking that in https://github.com/phetsims/models-of-the-hydrogen-atom/issues/74.

**Comment by terracoda** [comment]
**The Light Source interaction** is not being communicated as an on/off switch like in Molecules and Light. Are they not the same component?

**Electron Energy Levels** accordion title - "Levels" is plural in the PDOM, but singular on the screen. They should be the same, whichever one is correct.
-  hear  "Electron Energy Levels, empty region," after the help text. Is that intentional? I hear it even when the accordion is collapsed.

I have a proposed design for the Experiment - Model A-B Switch. I will create a separate issue for this. The on/off switch doesn't work for an A-B Switch.

**Absorption/ Emission** checkbox - @DianaTavares would it make sense to say "chart" or "table" in the name and/or help text instead of "dialog." Now that I have seen the content several times, the content looks like a single column or two column chart or table to me. I propose:
- help text: Show Absorption and Emission chart for electron state transitions and their associated wavelengths. 
- **Questions:** 
    - Is it a pedagogical problem that there is no "transition" information in Experiment mode?
    - How difficult is it to put the static content of the dialog in the sim, assuming the content is static?  Are the different symbols different types of photons?
    - And also wondering how easy it would be to add the the legend for proton, electron and photon to the PDOM? This goes beyond accessible names and help text, so just asking.
    
    Noting that sim crashed during this state with VO enabled - no interactions are functioning here:

![Screenshot 2024-11-06 at 15 34 28](https://github.com/user-attachments/assets/55419e33-1549-4f81-a8dc-a879a0fb6758)

**Atomic Model radiogroup on screen 1**
@DianaTavares, upon a closer look, I see "Classical" and "Quantum" flow for the models. Would it make sense to capture that representation in the help text?
![Screenshot 2024-11-06 at 15 42 01](https://github.com/user-attachments/assets/b3d8eaf6-02c3-422f-a5a2-5b339276ddd9)
- I propose changing: Choose atomic model to explore.
- To something like these: 
    - Choose from a range of classical and quantum models to explore.
    - Explore a range of classical and quantum models.

**Comment by terracoda** [comment]
Assigning to @DianaTavares and @pixelzoom to review suggested changes to help text. 

Any additional descriptions - the ones that go beyond accessible names and help text - would likely require a discussion with @kathy-phet. They may not be appropriate for the upcoming release.

**Comment by terracoda** [comment]
@pixelzoom, The change requests and additional comments are in [the  comment](https://github.com/phetsims/models-of-the-hydrogen-atom/issues/67#issuecomment-2460594348).

**Comment by pixelzoom** [comment]
@terracoda thanks for the feedback!

> The Light Source interaction is not being communicated as an on/off switch like in Molecules and Light. Are they not the same component?

They are not the same component. MOTHA uses scenery-phet's LaserPointerNode.  MAL has its own sim-specific component (no common code) called PhotonEmitterNode. It does some configuration of ARIA-related Property, but I can't tell why or what parts are important -- there are no comment details, and I can find no design paper trail.

> Electron Energy Levels accordion title - "Levels" is plural in the PDOM, but singular on the screen. They should be the same, whichever one is correct.

It should be singular, and was singular in the Java version -- I'll fix the PDOM and the design doc.  Singular is appropriate because hydrogen has 1 electron, the electron has 1 energy level at any given time, and the accordion box shows the enery level (state) that the electron is currently in.


> I have a proposed design for the Experiment - Model A-B Switch. I will create a separate issue for this. The on/off switch doesn't work for an A-B Switch.

No additional issue is needed. It’s being tracked in https://github.com/phetsims/models-of-the-hydrogen-atom/issues/74.


> Absorption/ Emission checkbox - [@DianaTavares](https://github.com/DianaTavares) would it make sense to say "chart" or "table" in the name and/or help text instead of "dialog." Now that I have seen the content several times, the content looks like a single column or two column chart or table to me. For help text, I propose: Show Absorption and Emission chart for electron state transitions and their associated wavelengths.

Defer to @[@DianaTavares](https://github.com/DianaTavares).  But might it be important for the user to know that in fact a Dialog is going to open displaying that chart/table?


> Is it a pedagogical problem that there is no "transition" information in Experiment mode?

No. It is a pedagogical design decision.


> How difficult is it to put the static content of the dialog in the sim, assuming the content is static? Are the different symbols different types of photons?

Quite difficult.  The static content is a large collection of buttons and text, and there’s no framework for coalescing that into “static content”.  I suspect that this woud be several hours of design and implementation work, involving multiple people.


> And also wondering how easy it would be to add the the legend for proton, electron and photon to the PDOM? This goes beyond accessible names and help text, so just asking.

First we’d need to define what "add to the PDOM" means.  Accessible name? Help text? Something else? The legend never has focus, so we’ve also need to decide whether this is appropriate to include in “Tier 1” description support. I suspect that this woud be several hours of design and implementation work, involving multiple people.  @kathy-phet do you want to expand what we’re doing for MOTHA to include this?


> Noting that sim crashed during this state with VO enabled - no interactions are functioning here:

Thanks for reporting. For future reference, when a sim crashes, it's helpful to open the browser console and report any details that are provided there. I'm guessing it hit https://github.com/phetsims/models-of-the-hydrogen-atom/issues/59, since that is a known problem in the Scrodinger model, "Experiment" uses the Schrodinger model, and you had the mode set to "Experiment".  So we don't need to be concerned about this crash for the purposes of this GitHub issue.


> [@DianaTavares](https://github.com/DianaTavares), upon a closer look, I see "Classical" and "Quantum" flow for the models. Would it make sense to capture that representation in the help text? …

This is related to the helpText of the radio button group used to choose the model.  The helpText is currently “Choose atomic model to explore.”  While there is a Classical/Quantum continuum in the Specta screen, there is no continuum in the Energy Levels screen — all of the models are quantum.  My understanding is that the current language was chosen to work for both screens.   So if you go down this path, we’ll need to customize the helpText per screen.   My recommendation is to leave it as is, but I’ll defer to [@DianaTavares](https://github.com/DianaTavares)

**Comment by pixelzoom** [comment]
11/7/24 design meeting: @DianaTavares @arouinfar @kathy-phet @terracoda @LindaStegemann @ariel-phet @KatieWoe @pixelzoom 

All feedback was discussed. Remaining work is being tracked in other GitHub issues.  So closing this issue.
